File: backend/src/main/java/com/eitek/incidents/application/services/AssigneeService.java
package com.eitek.incidents.application.services;

import com.eitek.incidents.application.dto.AssigneeDTO;
import com.eitek.incidents.domain.exceptions.InsufficientPermissionsException;
import com.eitek.incidents.domain.exceptions.ResourceNotFoundException;
import com.eitek.incidents.domain.models.Assignee;
import com.eitek.incidents.domain.models.AssigneeRoles;
import com.eitek.incidents.domain.repositories.AssigneeRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.stream.Collectors;

@Service
@RequiredArgsConstructor
public class AssigneeService {

    private final AssigneeRepository assigneeRepository;

    public AssigneeDTO createAssignee(AssigneeDTO assigneeDTO) {
        Assignee assignee = Assignee.builder()
                .name(assigneeDTO.getName())
                .role(assigneeDTO.getRole() != null ? assigneeDTO.getRole() : AssigneeRoles.OPERATOR)
                .isActive(true)
                .build();
        
        Assignee savedAssignee = assigneeRepository.save(assignee);
        return mapToDTO(savedAssignee);
    }

    public AssigneeDTO getAssigneeById(Long id) {
        Assignee assignee = assigneeRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("Assignee", "id", id));
        return mapToDTO(assignee);
    }

    public AssigneeDTO updateAssignee(Long id, AssigneeDTO assigneeDTO) {
        Assignee existingAssignee = assigneeRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("Assignee", "id", id));
        existingAssignee.setName(assigneeDTO.getName());
        if (assigneeDTO.getRole() != null) {
            existingAssignee.setRole(assigneeDTO.getRole());
        }
        Assignee updatedAssignee = assigneeRepository.save(existingAssignee);
        return mapToDTO(updatedAssignee);
    }

    public AssigneeDTO activateAssignee(Long id) {
        Assignee assignee = assigneeRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("Assignee", "id", id));
        assignee.activate();
        assignee.setUpdatedAt(java.time.LocalDateTime.now());
        Assignee savedAssignee = assigneeRepository.save(assignee);
        return mapToDTO(savedAssignee);
    }

    public AssigneeDTO deactivateAssignee(Long id) {
        Assignee assignee = assigneeRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("Assignee", "id", id));
        assignee.deactivate();
        assignee.setUpdatedAt(java.time.LocalDateTime.now());
        Assignee savedAssignee = assigneeRepository.save(assignee);
        return mapToDTO(savedAssignee);
    }

    public void deleteAssignee(Long id, Long deleterId) {
        Assignee assignee = assigneeRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("Assignee", "id", id));

        if (deleterId == null) {
            throw new ResourceNotFoundException("Deleter", "id", deleterId);
        }

        Assignee deleter = assigneeRepository.findById(deleterId)
                .orElseThrow(() -> new ResourceNotFoundException("Deleter", "id", deleterId));
        
        if (!deleter.getRole().equals(AssigneeRoles.ADMIN)) {
            throw new InsufficientPermissionsException(deleter.getName(), "delete assignees");
        }

        assigneeRepository.deleteById(id);
    }

    public List<AssigneeDTO> getAllAssignees() {
        return assigneeRepository.findAll().stream()
                .map(this::mapToDTO)
                .collect(Collectors.toList());
    }

    public List<AssigneeDTO> getActiveAssignees() {
        return assigneeRepository.findAll().stream()
                .filter(Assignee::isActive)
                .map(this::mapToDTO)
                .collect(Collectors.toList());
    }

    private AssigneeDTO mapToDTO(Assignee assignee) {
        return AssigneeDTO.builder()
                .id(assignee.getId())
                .name(assignee.getName())
                .role(assignee.getRole())
                .isActive(assignee.isActive())
                .createdAt(assignee.getCreatedAt())
                .updatedAt(assignee.getUpdatedAt())
                .build();
    }
}

File: frontend/src/application/services/AssigneeService.ts
import { Assignee } from "../../core/domain/Assignee";
import { AssigneeRoles } from "../../core/domain/AssigneeRoles";
import type { AssigneeDTO } from "../dto/AssigneeDTO";
import { AssigneeRepository } from "../../infrastructure/repositories/AssigneeRepository";

export class AssigneeService {
    private repository: AssigneeRepository;

    constructor() {
        this.repository = new AssigneeRepository();
    }

    async getAllAssignees(): Promise<Assignee[]> {
        const dtos = await this.repository.findAll();
        return dtos.map(dto => this.toDomain(dto));
    }

    async getActiveAssignees(): Promise<Assignee[]> {
        const dtos = await this.repository.findActive();
        return dtos.map(dto => this.toDomain(dto));
    }

    async getAssigneeById(id: number): Promise<Assignee> {
        const dto = await this.repository.findById(id);
        return this.toDomain(dto);
    }

    async createAssignee(data: { 
        name: string; 
        role: string;
    }): Promise<Assignee> {
        const dto = await this.repository.create({
            name: data.name,
            role: data.role
        });
        return this.toDomain(dto);
    }

    async updateAssignee(
        id: number, 
        data: { 
            name?: string; 
            role?: string;
        }
    ): Promise<Assignee> {
        const dto = await this.repository.update(id, {
            name: data.name,
            role: data.role
        });
        return this.toDomain(dto);
    }

    async activateAssignee(id: number): Promise<Assignee> {
        const dto = await this.repository.activate(id);
        return this.toDomain(dto);
    }

    async deactivateAssignee(id: number): Promise<Assignee> {
        const dto = await this.repository.deactivate(id);
        return this.toDomain(dto);
    }

    async deleteAssignee(id: number, userId: number): Promise<void> {
        await this.repository.delete(id, userId);
    }

    private toDomain(dto: AssigneeDTO): Assignee {
        return new Assignee({
            id: dto.id,
            name: dto.name,
            role: dto.role as AssigneeRoles,
            isActive: dto.isActive,
            createdAt: new Date(dto.createdAt),
            updatedAt: new Date(dto.updatedAt)
        });
    }

    toDTO(assignee: Assignee): AssigneeDTO {
        return {
            id: assignee.id,
            name: assignee.name,
            role: assignee.role,
            isActive: assignee.isActive,
            createdAt: assignee.createdAt.toISOString(),
            updatedAt: assignee.updatedAt.toISOString()
        };
    }
}